### PR TITLE
Fix #647 - Avoid store reading errors on mutations after a failed/canceled query

### LINKED ISSUE #647: Mutation will throw a missing store field if a query before errored or was canceled
This following line will throw: https://github.com/apollostack/apollo-client/blob/master/src/data/diffAgainstStore.ts#L339 on a mutation if a query before was canceled or errored. This is bad since it makes fail **every** mutation even the unrelated ones! I had hard time pinpointing this one but we should find an elegant solution to it. For now I'll try to set globally `throwOnMissingField` to false but I think we have ways to be smarter (preventing to register a failed or canceled result is one of those)

The stack I got is here:

```
Can't find field searchNodes({"query":"TEST","limit":40,"cursor":null}) on object ($ROOT_QUERY.connection({"id":"xxx"})) {
  "service": {
    "type": "id",
    "id": "$ROOT_QUERY.connection({\"id\":\"xxx\"}).service",
    "generated": true
  },
  "disconnected": false,
  "pins": [
    "$ROOT_QUERY.connection({\"id\":\"xxx\"}).pins.0",
    "$ROOT_QUERY.connection({\"id\":\"xxx\"}).pins.1",
    "$ROOT_QUERY.connection({\"id\":\"xxx\"}).pins.2",
    "$ROOT_QUERY.connection({\"id\":\"xxx\"}).pins.3"
  ],
  "user": {
    "type": "id",
    "id": "$ROOT_QUERY.connection({\"id\":\"xxx\"}).user",
    "generated": true
  },
  "id": "xxx",
  "handle": "xxx",
  "node({\"id\":\"root\"})": {
    "type": "id",
    "id": "$ROOT_QUERY.connection({\"id\":\"xxx\"}).node({\"id\":\"root\"})",
    "generated": true
  },
  "nodeChildren({\"id\":\"root\",\"limit\":40,\"cursor\":null,\"sortBy\":null})": {
    "type": "id",
    "id": "$ROOT_QUERY.connection({\"id\":\"xxx\"}).nodeChildren({\"id\":\"root\",\"limit\":40,\"cursor\":null,\"sortBy\":null})",
    "generated": true
  }
}.
Perhaps you want to use the `returnPartialData` option?"

Error
    at new ApolloError (webpack:///./~/apollo-client/errors.js?:14:22)
    at diffFieldAgainstStore (webpack:///./~/apollo-client/data/diffAgainstStore.js?:201:19)
    at eval (webpack:///./~/apollo-client/data/diffAgainstStore.js?:73:30)
    at Array.forEach (native)
    at diffSelectionSetAgainstStore (webpack:///./~/apollo-client/data/diffAgainstStore.js?:61:29)
    at diffFieldAgainstStore (webpack:///./~/apollo-client/data/diffAgainstStore.js?:256:16)
    at eval (webpack:///./~/apollo-client/data/diffAgainstStore.js?:73:30)
    at Array.forEach (native)
    at Object.diffSelectionSetAgainstStore (webpack:///./~/apollo-client/data/diffAgainstStore.js?:61:29)
    at Object.readSelectionSetFromStore (webpack:///./~/apollo-client/data/readFromStore.js?:31:37)
```

### PR DESCRIPTION
**Pull request by rricard**
Fixes #647 

TODO:

- [x] If this PR is a new feature, reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change
- [x] Add your name and email to the AUTHORS file (optional)
- [ ] Refactor

Post-mutation failing store reading during updateQueries.

### CODE REVIEW

**Comment by rricard** [comment]
I don't see how to pass AppVeyor here, something seems broken even before this PR, am I right ?

**Comment by rricard** [comment]
rebased, appveyor passes = yay!

**Comment by rricard** [comment]
rebased! @stubailo if you could consider a merge soon that would be great, this bug hurts us a lot lately. thanks!

**Comment by helfer** [comment]
@rricard can you work off of your own fork for now? I'd rather wait until we have a real fix than patch things up half way.

**Comment by helfer** [comment]
See also #487, which seems to be the same error.

**Comment by helfer** [comment]
Since this is a pretty serious bug imo, we'll make sure it gets fixed soon!

**Comment by helfer** [comment]
Also, how does this relate to #511, which you also filed?

**Comment by rricard** [comment]
@helfer whoops! forgot about this one! Closing it!

**Comment by rricard** [comment]
@helfer in general, working on a fork of the client is a huge pain, we did it once and we're not doing it again, I think (because you also need to fork react-apollo and last time we had to do that, we couldn't just get it out of git like this, we had to recompile + republish under a new name on npm...).

I really really prefer working on my own personal time on this to help you refactor all of this because the bug itself is really hard on us (cancellation is the biggest issue I think, fails after errored queries just made us make the schema more stable, which was a good thing).

**Comment by rricard** [comment]
In general, I agree, that's not the kind of fix I like too, for me the idea was more about providing a test and an early fix. Then we refactor that ASAP but with less pressure since the bug is not there, the test maintaining it down. I think, as discussed in #615, that by simplifying the query diffing we can get it right.

**Comment by rricard** [comment]
Also the solution you are proposing seems good, I didn't thought about offline support and I have to say that's a complex problem. I'll see tomorrow if I can get a better fix without refactoring the whole thing!

File: src/QueryManager.ts [hunk 1]
```diff
@@ -631,17 +631,24 @@ export class QueryManager {
...
+        rootId: 'ROOT_QUERY',
+        selectionSet: queryDefinition.selectionSet,
+        variables: queryOptions.variables,
+        returnPartialData: queryOptions.returnPartialData || queryOptions.noFetch,
+        fragmentMap: createFragmentMap(fragments || []),
+      });
+    } catch (e) {
+      if (!/can't find field/i.test(e.message)) {
+        throw e;
+      }
```
Comment (rricard) [inline src/QueryManager.ts]: @helfer I agree, this feels completely hacky now that I'm coming back here

**Comment by rricard** [comment]
@helfer or anyone affected:

I'm getting the patched version like this in the meantime, not perfect, but will do the trick:

`package.json
```json
{
  "scripts": {
    "postinstall": "cd ./node_modules/apollo-client; npm i; npm run prepublish"
  },
  "dependencies": {
    "apollo-client": "apollostack/apollo-client#fix-647"
  }
}
```

Note that doing this will slow down pretty much everything... (CI, deploy, etc...)

**Comment by rricard** [comment]
And... no it doesn't work, tsconfig gets npmignored

**Comment by rricard** [comment]
Removing the NPM ignore for now so I can use it as a fork!

**Comment by helfer** [comment]
@rricard ok, I'll try to fix this soon. It might take until next week though, because we'll be away from tomorrow until Monday.

**Comment by rricard** [comment]
@helfer don't worry, either I'll find some time to propose a better fix, either I'll wait, take your time away, enjoy it and don't think about this bug! 👍

**Comment by rricard** [comment]
Closing this one: followup in #696

**Pull request closed without merging**